File: src/api/routers/foundation/evidence.py
"""Read-only Audit Evidence API — Rule 71 §6.

Domain exceptions are not caught here — `EXCEPTION_MAP` in
`src/api/contracts/exception_mapping.py` wraps and translates them via the
global handler (§9 PLT-21 decision, task-1108).
"""
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, Query

from src.api.contracts.envelope import ApiResponse, ok
from src.api.deps import get_current_admin, get_current_user
from src.api.foundation_deps import get_audit_event_repository
from src.api.schemas.foundation.evidence import AuditTimelinePage
from src.foundation.evidence.application.get_audit_timeline import get_audit_timeline
from src.foundation.evidence.application.verify_audit_chain import verify_audit_chain
from src.foundation.evidence.ports.repository import AuditEventRepository
from src.services.auth_service import User

router = APIRouter(prefix="/v1/foundation/evidence", tags=["foundation:evidence"])


@router.get("/timeline")
async def get_timeline(
    cursor: str | None = None,
    limit: int = Query(default=50, ge=1, le=100),
    aggregate_type: str | None = None,
    action: str | None = None,
    user: User = Depends(get_current_user),
    repo: AuditEventRepository = Depends(get_audit_event_repository),
) -> ApiResponse[AuditTimelinePage]:
    page = await get_audit_timeline(
        repo,
        tenant_id=user.user_id,
        cursor=cursor,
        limit=limit,
        aggregate_type=aggregate_type,
        action=action,
    )
    return ok(page)


@router.post("/chain:verify")
async def post_verify_chain(
    tenant_id: UUID | None = Query(default=None),
    admin: User = Depends(get_current_admin),
    repo: AuditEventRepository = Depends(get_audit_event_repository),
) -> ApiResponse[dict[str, bool]]:
    """AUD-003 operational tool — Rule 79 §4. Admin-only (not exposed to
    regular users since it can query the full chain or a specific tenant).
    Omitting `tenant_id` verifies only the system event chain (tenant_id IS
    NULL) — iterating all tenants is out of scope for this leaf (operational
    batch task target)."""
    await verify_audit_chain(repo, tenant_id)
    return ok({"verified": True})

File: frontend/packages/api-client/src/clients/evidence.ts
import type { AuditTimelinePage, ChainVerificationResult } from "@aios/shared-types";
import { resolveEnvelope, resolvePath } from "../apiPaths";
import type { AnyConstructor } from "../http";

export interface AuditTimelineParams {
  cursor?: string;
  limit?: number;
  aggregateType?: string;
  action?: string;
}

// task-2337(FE-OPS-3): 증빙(evidence) 타임라인 조회 + 감사 체인 검증(chain:verify)
// 클라이언트 — src/api/routers/foundation/evidence.py 원문 기준. 둘 다
// `-> ApiResponse[...]`+`ok(...)`로 응답한다(envelope=true). chain:verify는
// tenant_id를 생략하면 서버가 system 이벤트 체인만 검증한다(admin.py:41-48
// 원문) — 여기서는 값을 그대로 옮기고 판정(해시 체인 검증)은 재구현하지
// 않는다(decision: LC-3 hash_chain 재구현 금지).
export function withEvidence<TBase extends AnyConstructor>(Base: TBase) {
  return class extends Base {
    async getAuditTimeline(params: AuditTimelineParams = {}): Promise<AuditTimelinePage> {
      const path = this.withQuery(resolvePath("evidence.timeline"), {
        cursor: params.cursor,
        limit: params.limit,
        aggregate_type: params.aggregateType,
        action: params.action,
      });
      return resolveEnvelope("evidence.timeline") ? this.requestEnvelope(path) : this.request(path);
    }

    async verifyAuditChain(tenantId?: string): Promise<ChainVerificationResult> {
      const path = this.withQuery(resolvePath("evidence.chainVerify"), { tenant_id: tenantId });
      return this.postEnvelope(path);
    }
  };
}
